sockets.js

```diff
@@ -443,48 +487,53 @@ if (cluster.isMaster) {
 		sockets.clear();
 		channels.clear();
 		subchannels.clear();
+
 		app.close();
 		if (appssl) appssl.close();
+
+		// Let the server(s) finish closing.
+		setImmediate(() => process.exit(0));
 	});
 
 	// this is global so it can be hotpatched if necessary
 	let isTrustedProxyIp = Dnsbl.checker(Config.proxyip);
 	let socketCounter = 0;
 	server.on('connection', socket => {
-		if (!socket) {
-			// For reasons that are not entirely clear, SockJS sometimes triggers
-			// this event with a null `socket` argument.
-			return;
-		} else if (!socket.remoteAddress) {
-			// This condition occurs several times per day. It may be a SockJS bug.
+		// For reasons that are not entirely clear, SockJS sometimes triggers
+		// this event with a null `socket` argument.
+		if (!socket) return;
+
+		if (!socket.remoteAddress) {
+			// SockJS sometimes fails to be able to cache the IP, port, and
+			// address from connection request headers.
 			try {
 				socket.destroy();
 			} catch (e) {}
 			return;
 		}
 
-		let socketid = socket.id = '' + (++socketCounter);
+		let socketid = '' + (++socketCounter);
 		sockets.set(socketid, socket);
 
-		if (isTrustedProxyIp(socket.remoteAddress)) {
+		let socketip = socket.remoteAddress;
+		if (isTrustedProxyIp(socketip)) {
 			let ips = (socket.headers['x-forwarded-for'] || '').split(',');
-			let ip;
-			while ((ip = ips.pop())) {
-				ip = ip.trim();
-				if (!isTrustedProxyIp(ip)) {
-					socket.remoteAddress = ip;
+			for (let i = ips.length; i--;) {
```

Comment:
What is this poorly optimized reverse iteration?

`for (ip of ips) {`

Reply:
Would needing to reverse the IPs array in order to do that still make it more optimal?

Reply:
Functions with reverse loops seem to be able to be optimized by Turbofan. Is there a different reason it'd be poorly optimized?
```javascript
> const foo = () => { let bar = [5,4,3,2,1,0]; for (let i = bar.length; i--;) { let baz = bar[i]; } }                   undefined
> foo(); foo(); foo(); %OptimizeFunctionOnNextCall(foo); foo(); %GetOptimizationStatus(foo)                             49
> 49 >> 7
0
```

Reply:
And yeah, I know it's dumb to argue over this when the size of the array's so small, but I do use them when I don't have a good enough reason to reverse the array then loop through it

Reply:
Is there any reason this needs to be looped backwards?

If so, I'd at least do

`for (let i = ips.length; i > 0; i--) {`

for readability

Reply:
I hhink it does, since looping forwards might not give the most accurate hostname in the list if there are multiple IPs in the list that'd get matched as proxies